src/engine/DnCWorker.cpp

```diff
@@ -60,132 +60,122 @@ void DnCWorker::setQueryDivider( DivideStrategy divideStrategy )
-            // object of class DnCStatistics, which contains some basic
-            // statistics. The maps are owned by the DnCManager.
+        // Reset the engine state
+        _engine->restoreState( *_initialState );
+        _engine->reset();
 
-            // Apply the split and solve
-            _engine->applySplit( *split );
-            _engine->solve( timeoutInSeconds );
+        // TODO: each worker is going to keep a map from *CaseSplit to an
+        // object of class DnCStatistics, which contains some basic
+        // statistics. The maps are owned by the DnCManager.
 
-            Engine::ExitCode result = _engine->getExitCode();
-            printProgress( queryId, result );
-            // Switch on the result
-            if ( result == Engine::UNSAT )
-            {
-                // If UNSAT, continue to solve
-                *_numUnsolvedSubQueries -= 1;
-                delete subQuery;
-            }
-            else if ( result == Engine::TIMEOUT )
+        // Apply the split and solve
+        _engine->applySplit( *split );
+        _engine->solve( timeoutInSeconds );
+
+        IEngine::ExitCode result = _engine->getExitCode();
+        printProgress( queryId, result );
+        // Switch on the result
+        if ( result == IEngine::UNSAT )
+        {
+            // If UNSAT, continue to solve
+            *_numUnsolvedSubQueries -= 1;
+            if ( _numUnsolvedSubQueries->load() == 0 )
+                *_shouldQuitSolving = true;
+            delete subQuery;
+        }
+        else if ( result == IEngine::TIMEOUT )
+        {
+            // If TIMEOUT, split the current input region and add the
+            // new subQueries to the current queue
+            SubQueries subQueries;
+            _queryDivider->createSubQueries( pow( 2, _onlineDivides ),
+                                             queryId, *split,
+                                             (unsigned) timeoutInSeconds *
+                                             _timeoutFactor, subQueries );
+            for ( auto &newSubQuery : subQueries )
             {
-                // If TIMEOUT, split the current input region and add the
-                // new subQueries to the current queue
-                SubQueries subQueries;
-                _queryDivider->createSubQueries( pow( 2, _onlineDivides ),
-                                                 queryId, *split,
-                                                 (unsigned) timeoutInSeconds *
-                                                 _timeoutFactor, subQueries );
-                for ( auto &newSubQuery : subQueries )
+                if ( !_workload->push( std::move( newSubQuery ) ) )
                 {
-                    if ( !_workload->push( std::move( newSubQuery ) ) )
-                    {
-                        ASSERT( false );
-                    }
-
-                    *_numUnsolvedSubQueries += 1;
+                    ASSERT( false );
```

Comment:
Should we perhaps throw an error here? In non-DEBUG mode, the execution won't stop here.

Reply:
Throwing error here.